test/export/test_passes.py

```diff
@@ -58,6 +68,109 @@ def count_additions(gm) -> int:
+
+    def test_runtime_assert_multiple_dims(self) -> None:
+        class M(torch.nn.Module):
+            def __init__(self):
+                super().__init__()
+
+            def forward(self, x, y):
+                return x.cos().sum() + y.sin().sum()
+
+        x = torch.zeros(4, 2, 3)
+        y = torch.zeros(5, 5, 5)
+
+        constraints = [
+            dynamic_dim(x, 1) >= 2,
+            dynamic_dim(x, 1) <= 6,
+            dynamic_dim(y, 0) >= 3,
+            dynamic_dim(x, 0) >= 3
+        ]
+
+        gm = _export(M(), (x, y), constraints=constraints)
+
+        pass_result = AddRuntimeAssertionsForConstraintsPass()(gm)
+        self.assertTrue(pass_result.modified)
+
+        num_assert = count_call_function(pass_result.graph_module.graph, torch.ops.aten._assert_async.msg)
+        num_scalar_tensor = count_call_function(pass_result.graph_module.graph, torch.ops.aten.scalar_tensor.default)
+
+        self.assertEqual(num_assert, 6)
+        self.assertEqual(num_scalar_tensor, 6)
+
+        with self.assertRaisesRegex(RuntimeError, "Input #0"):
+            pass_result.graph_module(torch.zeros(4, 7, 3), torch.ones(5, 5, 5))
+
+        with self.assertRaisesRegex(RuntimeError, "Input #1"):
+            pass_result.graph_module(torch.zeros(4, 2, 3), torch.ones(2, 5, 5))
+
+    def test_runtime_assert_some_dims_not_specified(self) -> None:
+        class M(torch.nn.Module):
+            def __init__(self):
+                super().__init__()
+
+            def forward(self, x, y):
+                return x.cos().sum() + y.sin().sum()
+
+        x = torch.zeros(4, 2, 3)
+        y = torch.zeros(5, 5, 5)
+
+        constraints = [
+            dynamic_dim(x, 1) >= 2,
+            dynamic_dim(x, 1) <= 6,
+            dynamic_dim(x, 0) >= 3
+        ]
+
+        gm = _export(M(), (x, y), constraints=constraints)
+
+        pass_result = AddRuntimeAssertionsForConstraintsPass()(gm)
+        self.assertTrue(pass_result.modified)
+
+        num_assert = count_call_function(pass_result.graph_module.graph, torch.ops.aten._assert_async.msg)
+        num_scalar_tensor = count_call_function(pass_result.graph_module.graph, torch.ops.aten.scalar_tensor.default)
+
+        # there are 3 asserts from y and 2 from dynamic x dims and 1 from static x dim
+        self.assertEqual(num_assert, 6)
+        self.assertEqual(num_scalar_tensor, 6)
+
+        with self.assertRaisesRegex(RuntimeError, "Input #0"):
+            pass_result.graph_module(torch.zeros(4, 7, 3), torch.ones(5, 5, 5))
+
+        # y is specialized to 5
+        with self.assertRaisesRegex(RuntimeError, "Input #1's dimension #0 size is specialized at 5"):
+            pass_result.graph_module(torch.zeros(4, 2, 3), torch.ones(2, 5, 5))
+
+        # Since we didn't insert the constraint for x[1] >= 2, it should work for case where x[1] == 1
+        gm_result_for_1_size = pass_result.graph_module(torch.ones(3, 1, 3), torch.ones(5, 5, 5))
```

Comment:
I think the contract is that export API will guarantee that user specified constraints are always respected (as long as valid). This seems like a flaw and it will be a surprise to users when `dynamic_dim(x, 1) >= 2` is specified but not enforced. I think the motivation of not insert assertion for `x[1] >= 2` is because of 0/1 specialization where in most cases, N >= 2 will work also work for N == 1. Is it the reason?

If yes maybe a better option is to let the compiler to figure out whether the exported graph at dim 1 can be correctly generalize to N == 1, which should be simple and clean to validate in a separate pass prior to insert runtime assertion. If N == 1 works correctly, remove constraint for `dynamic_dim(x, 1) >= 2` and warn users, or error out and let users remove the constraint from the input explicitly, the remaining input constraints will be correct and necessary.

Reply:
Yeah our assumption is that if we were able to get a graph for N >= 2, it should work for cases that N == 1. 
For the validation, I agree that we need an infra that can check if our graph can be correctly specialized for N==1 cases. Do you think we can do it in a follow up diff? 
cc: @avikchaudhuri

Reply:
Yeah not mean to block. Add a comment/todo about the assumption, and follow up in a separate PR is okay.

The important thing is that I want to make sure this will be handled decently as part of export contract. Currently validating whether the graph can be correctly generalized to N == 1 or not is user's responsibility, e.g. they have to write tests for it. This is conflict to not exposing 0/1 specialization to end users, and what's worse when users realize that the exported graph can not be generalized to N == 1, they have no way to enforce the correctness because the constraint N >= 2 is not installed and there are no knobs for users to correct it manually. So IMO it's best to make it part of export contract. 

Like to hear your thoughts @avikchaudhuri

Reply:
As discussed in the meeting yesterday, there are two cases:
* If export passes with N >= 2, we only have a soft expectation that it should also work with N == 1. This can only be covered by testing, I don't think we can validate this automatically in general. It also feels wrong to disallow N == 1 in that case. So I suggest we just ask people to test if they think that N == 1 should be a valid input.
* If you try exporting with N == 1, we should add a `log.warning` that we are implicitly specializing. If you marked that dimension dynamic, we should error.

Reply:
@avikchaudhuri I agree with #2 _(warning when exporting with N == 1)_, I think #1 needs more discussion.

> we only have a soft expectation that it should also work with N == 1

That's the problem because soft expectation means no guarantee at all and hope it works. When it doesn't work, it will come as a surprise to users, as they have specified `dynamic_dim(x, 1) >= 2` as an input constraint based on the suggestion provided by the compiler! The promise of user specified constraints will be respected faithfully is no longer true in this case.

> This can only be covered by testing, I don't think we can validate this automatically in general.

I fail to understand why this can not be validated automatically in general if this is testable by users. Maybe you could help elaborate why you think the approach I proposed in previous comment may not work? We can brainstorm other solutions when we agree this is an issue we need to take care for the users.
 
 > It also feels wrong to disallow N == 1 in that case. So I suggest we just ask people to test if they think that N == 1 should be a valid input.
 
I'm not suggesting whether N == 1 should be allowed or not allowed. If we can reason about either way I will be fine. My point is that the contract provided by export should be cleaner and stronger, at least on input shape constraints.

Reply:
OK, so we are only talking about you passing in an example input where N >= 2 and we pass, right? Now we have only two options:
1. Assert that N >= 2. I don't like this option because in practice your code might be fine with N == 1, and you can't export with N == 1 and dynamic, so you're stuck if you also wanted N == 1 to work.
2. Automatically validate. This seems tricky. Will we come up with new example inputs with N == 1? Or basically use fake tensors directly? Will we try all combinatorial possibilities for all dynamic inputs, setting each N to 1? Etc. It is because of this trickiness that I want to just suggest that users do testing for this case.

Does this make sense? Please let me know if I'm missing anything easy here.

> When it doesn't work, it will come as a surprise to users, as they have specified dynamic_dim(x, 1) >= 2 as an input constraint based on the suggestion provided by the compiler! The promise of user specified constraints will be respected faithfully is no longer true in this case.

I didn't understand this. If you specify dynamic_dim(x, 1) >= 2, and if it doesn't work with N == 1 then it's fine! So we're only talking about when you specify dynamic_dim(x, 1) without bounds.

Reply:
@avikchaudhuri Sorry for late reply. I want to emphasize that my point is that I feel the export contract should be stronger with 0/1 specialization.

> I didn't understand this. If you specify dynamic_dim(x, 1) >= 2, and if it doesn't work with N == 1 then it's fine!

It's fine only if it happens to work for N == 1 because users won't get the wrong result. It's not fine if it doesn't work for `N == 1`. I think this is the key point we haven't aligned yet. 
For example, say I have `2 <= dynamic_dim(x, 1) <= 10`, if I give an input `x.size(1) == 11`, it errors out immediately right? Even if `x.size(1) == 11` may happen to work, we want to error out anyway because that's the constraint user specified. Why is `x.size(1) == 1` treated differently? 


> Now we have only two options:

I don't see option 1 & 2 are exclusive. We probably should assert `N >= 2` conditionally. So let's talk about option 2 first. If there is a way to validate whether or not the exported graph (for `N >= 2` at all dynamic dims) works for `N == 1`, then we can assert `N >= 2` conditionally. Actually there is not much dims to test against. It's linear. If you have an input tensor of size `(x, y, z)`, in the worst case where all dims are dynamic, you generate `x + y + z` inputs in total that toggle one dim to 1 a time to test if the graph can be correctly generalized to N == 1, you only install assertion of `N >= 2` on dim where `N == 1` produces inconsistent output.

Reply:
> For example, say I have 2 <= dynamic_dim(x, 1) <= 10, if I give an input x.size(1) == 11, it errors out immediately right? Even if x.size(1) == 11 may happen to work, we want to error out anyway because that's the constraint user specified. Why is x.size(1) == 1 treated differently?

It is treated differently because PyTorch treats it differently. I'm assuming there is a good reason for 0/1 specialization: I'm guessing it is because many torch operators behave differently for 0/1 than others. This is something I'm willing to defer to and explain away based on general PyTorch semantics.

**Is it ideal? NO. So I fully agree with you on the jankiness here.** But at the same time I'm not comfortable challenging the assumption that 0/1 specialization predates export and exists for a good reason, and taking a hard stand with export without understanding all the reasoning behind why it exists.

> Actually there is not much dims to test against. It's linear.

This is an incomplete test. For each dynamic, we have two cases, 1 or >= 2. So a complete test is exponential: 2^N. You're missing the case where x = 1, y = 1, z >= 2, e.g.

I won't oppose doing the incomplete linear test to surface additional signal.